src/java.base/share/classes/java/lang/invoke/LazyInitializingVarHandle.java

```diff
@@ -0,0 +1,137 @@
+
+import static java.lang.invoke.MethodHandleStatics.UNSAFE;
+import static java.lang.invoke.MethodHandleStatics.uncaughtException;
+import static java.lang.invoke.MethodHandles.Lookup.IMPL_LOOKUP;
+
+/**
+ * A lazy initializing var handle. It lazily initializes the referenced class before
+ * any invocation of the target var handle to prevent reading uninitialized static
+ * final field values.
+ */
+final class LazyInitializingVarHandle extends VarHandle {
+
+    private static final MethodHandle MH_ensureInitialized;
+    private final VarHandle target;
+    private final Class<?> refc;
+    private @Stable boolean initialized;
+
+    static {
+        try {
+            MH_ensureInitialized = IMPL_LOOKUP.findVirtual(LazyInitializingVarHandle.class, "ensureInitialized",
+                    MethodType.methodType(void.class));
+        } catch (Throwable ex) {
+            throw uncaughtException(ex);
+        }
+    }
+
+    LazyInitializingVarHandle(VarHandle target, Class<?> refc) {
+        super(target.vform, target.exact);
+        this.target = target;
+        this.refc = refc;
+    }
+
+    @Override
+    MethodType accessModeTypeUncached(AccessType at) {
+        return target.accessModeTypeUncached(at);
+    }
+
+    @Override
+    VarHandle asDirect() {
+        return target;
+    }
+
+    @Override
+    @ForceInline
+    boolean checkAccessModeThenIsDirect(AccessDescriptor ad) {
+        super.checkAccessModeThenIsDirect(ad);
+        return false;
+    }
+
+    @Override
+    public VarHandle withInvokeExactBehavior() {
+        if (!initialized && hasInvokeExactBehavior())
+            return this;
+        var updatedDelegate = target.withInvokeExactBehavior();
+        return initialized ? updatedDelegate : new LazyInitializingVarHandle(updatedDelegate, refc);
+    }
+
+    @Override
+    public VarHandle withInvokeBehavior() {
+        if (!initialized && !hasInvokeExactBehavior())
+            return this;
+        var updatedDelegate = target.withInvokeBehavior();
+        return initialized ? updatedDelegate : new LazyInitializingVarHandle(updatedDelegate, refc);
+    }
+
+    @Override
+    public Optional<VarHandleDesc> describeConstable() {
+        return target.describeConstable();
+    }
+
+    private void ensureInitialized() {
+        if (this.initialized)
+            return;
+
+        UNSAFE.ensureClassInitialized(refc);
+        this.initialized = true;
+
+        var cache = this.methodHandleTable;
+        int len = cache.length;
+        for (int i = 0; i < len; i++) {
+            var mh = cache[i];
+            if (mh != null) {
+                var callTarget = target.getMethodHandle(i);
+                mh.updateForm(new Function<>() {
+                    @Override
+                    public LambdaForm apply(LambdaForm lambdaForm) {
+                        return callTarget.form;
+                    }
+                });
+                cache[i] = callTarget;
+            }
+        }
```

Comment:
Unlike calling `Unsafe::ensureClassInitialized(…)` (which performs its own synchronisation) or writing `true` to `this.initialized` (which is a benign race), updating the lambda forms this way is probably not quite correct to do under a race and probably needs some synchronisation.

Reply:
The form update is a benign race. The array update might need some fences, but I'm not sure, as the read MH should be complete when the updated value is already observed.